pkg/container/vector/vector.go

```diff
@@ -72,1608 +65,757 @@ func (v *Vector) GetIsBin() bool {
 }
 
-// XXX Compatibility: how many aliases do we need ...
-func GetStrColumn(v *Vector) []string {
-	return GetStrVectorValues(v)
+func (v *Vector) IsConst() bool {
+	return v.class == CONSTANT
 }
 
-// Get Value at index
-func GetValueAt[T types.FixedSizeT](v *Vector, idx int64) T {
-	return MustTCols[T](v)[idx]
+// IsConstNull return true if the vector means a scalar Null.
+// e.g.
+//
+//	a + Null, and the vector of right part will return true
+func (v *Vector) IsConstNull() bool {
+	return v.IsConst() && v.nsp != nil && nulls.Contains(v.nsp, 0)
 }
 
-func GetValueAtOrZero[T types.FixedSizeT](v *Vector, idx int64) T {
-	var zt T
-	ts := MustTCols[T](v)
-	if int64(len(ts)) <= idx {
-		return zt
+func (v *Vector) Free(m *mpool.MPool) {
+	// const vector's data & area allocate with nil,
+	// so we can't free it by using mpool.
+	if v.col != nil {
+		m.Free((*(*[]byte)(unsafe.Pointer(v.col))))
 	}
-	return ts[idx]
-}
-
-// Get the pointer to idx-th fixed size entry.
-func GetPtrAt(v *Vector, idx int64) unsafe.Pointer {
-	return unsafe.Pointer(&v.data[idx*int64(v.GetType().TypeSize())])
-}
-
-// Raw version, get from v.data.   Adopt python convention and
-// neg idx means counting from end, that is, -1 means last element.
-func (v *Vector) getRawValueAt(idx int64) []byte {
-	tlen := int64(v.GetType().TypeSize())
-	dlen := int64(len(v.data))
-	if idx >= 0 {
-		if dlen < (idx+1)*tlen {
-			panic("vector invalid index access")
-		}
-		return v.data[idx*tlen : idx*tlen+tlen]
-	} else {
-		start := dlen + tlen*idx
-		end := start + tlen
-		if start < 0 {
-			panic("vector invalid index access")
-		}
-		return v.data[start:end]
+	if v.area != nil {
+		m.Free(v.area)
 	}
+	v.col = nil
+	v.area = nil
 }
 
 func (v *Vector) MarshalBinary() ([]byte, error) {
 	var buf bytes.Buffer
 
-	if v.isConst {
-		i64 := int64(v.ScalarLength())
-		buf.WriteByte(1)
-		buf.Write(types.EncodeInt64(&i64))
-	} else {
-		buf.WriteByte(0)
-		// length, even not used, let's fill it.
-		i64 := int64(0)
-		buf.Write(types.EncodeInt64(&i64))
+	buf.WriteByte(uint8(v.class))
+	{ // write length
+		length := int64(v.length)
+		buf.Write(types.EncodeInt64(&length))
 	}
-	data, err := v.Show()
-	if err != nil {
-		return nil, err
+	{ // write type
+		data := types.EncodeType(&v.typ)
+		buf.Write(data)
+	}
+	{ // write nspLen, nsp
+		data, err := v.nsp.Show()
+		if err != nil {
+			return nil, err
+		}
+		length := uint32(len(data))
+		buf.Write(types.EncodeUint32(&length))
+		if len(data) > 0 {
+			buf.Write(data)
+		}
+	}
+	{ // write colLen, col
+		data := (*(*[]byte)(unsafe.Pointer(v.col)))
+		length := uint32(v.length * v.typ.TypeSize())
+		if len(data[:length]) > 0 {
+			buf.Write(data[:length])
+		}
+	}
+	{ // write areaLen, area
+		length := uint32(len(v.area))
+		buf.Write(types.EncodeUint32(&length))
+		if len(v.area) > 0 {
+			buf.Write(v.area)
+		}
 	}
-	buf.Write(data)
 	return buf.Bytes(), nil
 }
 
 func (v *Vector) UnmarshalBinary(data []byte) error {
```

Comment:
Is there a way to mark a vector readonly?   after this call, I think your vector is readonly, am i right?

Reply:
We should get rid of gob.   Can we just get rid of this method and whoever use gob.